base/applications/cmdutils/where/where.c

```diff
@@ -0,0 +1,486 @@
+static strlist_t s_patterns = strlist_default;
+static strlist_t s_results = strlist_default;
+static strlist_t s_pathext = strlist_default;
+
+// is it either "." or ".."?
+#define IS_DOTS(pch) \
+    (*(pch) == L'.' && ((pch)[1] == 0 || ((pch)[1] == L'.' && (pch)[2] == 0)))
+
+#define DEFAULT_PATHEXT L".com;.exe;.bat;.cmd"
+
+typedef enum WRET // return code of WHERE command
+{
+    WRET_SUCCESS = 0,
+    WRET_NOT_FOUND = 1,
+    WRET_ERROR = 2
+} WRET;
+
+static VOID WhereError(UINT nID)
+{
+    if (!(s_dwFlags & FLAG_Q)) // not quiet mode?
+        ConResPuts(StdErr, nID);
+}
+
+typedef BOOL (CALLBACK *WHERE_CALLBACK)(LPCWSTR pattern, LPCWSTR path, PWIN32_FIND_DATAW data);
+
+static BOOL
+WhereSearchGeneric(LPCWSTR pattern, LPWSTR path, size_t path_len, BOOL bDir,
+                   WHERE_CALLBACK callback)
+{
+    LPWSTR pch;
+    size_t cch;
+    BOOL ret;
+    WIN32_FIND_DATAW data;
+    HANDLE hFind = FindFirstFileExW(path, FindExInfoStandard, &data, FindExSearchNameMatch,
+                                    NULL, 0);
+    if (hFind == INVALID_HANDLE_VALUE)
+        return TRUE; // not found
+
+    pch = wcsrchr(path, L'\\') + 1;
+    cch = path_len - (pch - path);
+    do
+    {
+        if (bDir != !!(data.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY))
+            continue;
+        if (bDir && IS_DOTS(data.cFileName))
+            continue; // ignore "." and ".."
+        if (data.dwFileAttributes & FILE_ATTRIBUTE_VIRTUAL)
+            continue; // ignore virtual
+
+        StringCchCopyW(pch, cch, data.cFileName); // build full path
+
+        ret = callback(pattern, path, &data);
+        if (!ret) // out of memory
+            break;
+    } while (FindNextFileW(hFind, &data));
+    FindClose(hFind);
+    return ret;
+}
+
+static BOOL CALLBACK WherePrintPath(LPCWSTR pattern, LPCWSTR path, PWIN32_FIND_DATAW data)
+{
+    WCHAR szPath[MAX_PATH + 2], szDate[32], szTime[32];
+    LARGE_INTEGER FileSize;
+    FILETIME ftLocal;
+    SYSTEMTIME st;
+
+    if (strlist_find_i(&s_results, path) >= 0)
+        return TRUE; // already exists
+    if (!strlist_add(&s_results, path))
+        return FALSE; // out of memory
+    if (s_dwFlags & FLAG_Q) // quiet mode?
+        return TRUE;
+
+    if (s_dwFlags & FLAG_T) // print detailed info
+    {
+        // convert date/time
+        FileTimeToLocalFileTime(&data->ftLastWriteTime, &ftLocal);
+        FileTimeToSystemTime(&ftLocal, &st);
+        // get date/time strings
+        GetDateFormatW(LOCALE_USER_DEFAULT, 0, &st, NULL, szDate, _countof(szDate));
+        GetTimeFormatW(LOCALE_USER_DEFAULT, 0, &st, NULL, szTime, _countof(szTime));
+        // set size
+        FileSize.LowPart = data->nFileSizeLow;
+        FileSize.HighPart = data->nFileSizeHigh;
+        // print
+        if (s_dwFlags & FLAG_F) // double quote
+            StringCbPrintfW(szPath, sizeof(szPath), L"\"%s\"", path);
+        else
+            StringCbCopyW(szPath, sizeof(szPath), path);
```

Comment:
```suggestion
            StringCchCopyW(szPath, _countof(szPath), path);
```

Reply:
(So that the code uniformly only uses StringCch* functions)

Reply:
OK.